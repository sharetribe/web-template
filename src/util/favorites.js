import { createResourceLocatorString } from './routes';

export const handleToggleFavorites = parameters => async isFavorite => {
  const { onFetchCurrentUser, routes, location, history } = parameters;
  const currentUser = await onFetchCurrentUser();
  // Only allow signed-in users to save favorites
  if (!currentUser || Object.keys(currentUser).length === 0) {
    const state = {
      from: `${location.pathname}${location.search}${location.hash}`,
    };
    // Sign up and return back to the listing page.
    history.push(createResourceLocatorString('SignupPage', routes, {}, {}), state);
  } else {
    const { listingId, listingType, onUpdateFavorites } = parameters;
    const {
      attributes: { profile },
    } = currentUser;
    const { privateData } = profile;
    const favorites = privateData?.favorites;
    const categoryFavorites = favorites?.[listingType];
    const noPrivateData = !privateData;
    const noFavorites = !favorites;
    const noFavoritesCategory = !categoryFavorites;
    let payload;
    if (noPrivateData || noFavorites) {
      payload = {
        privateData: {
          favorites: {
            [listingType]: [listingId],
          },
        },
      };
    } else if (noFavoritesCategory) {
      payload = {
        privateData: {
          favorites: {
            ...favorites,
            [listingType]: [listingId],
          },
        },
      };
    } else {
      if (isFavorite) {
        payload = {
          privateData: {
            favorites: {
              ...favorites,
              [listingType]: categoryFavorites.filter(f => f !== listingId),
            },
          },
        };
      } else {
        payload = {
          privateData: {
            favorites: {
              ...favorites,
              [listingType]: [...categoryFavorites, listingId],
            },
          },
        };
      }
    }
    onUpdateFavorites(payload);
  }
};
