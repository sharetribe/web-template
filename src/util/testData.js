import Decimal from 'decimal.js';
import { types as sdkTypes } from './sdkLoader';
import { daysBetween, getStartOf } from './dates';
import { LISTING_STATE_PUBLISHED, TIME_SLOT_TIME } from './types';
import { getProcess, TX_TRANSITION_ACTOR_CUSTOMER } from '../transactions/transaction';

const { UUID, LatLng, Money } = sdkTypes;
const processTransitions = getProcess('default-purchase')?.transitions;

// Create a booking that conforms to the util/types booking schema
export const createBooking = (id, attributes = {}) => {
  const start = attributes.start || new Date(Date.UTC(2017, 5, 10));
  const end = attributes.end || new Date(Date.UTC(2017, 5, 10));
  return {
    id: new UUID(id),
    type: 'booking',
    attributes: {
      start,
      displayStart: start,
      end,
      displayEnd: end,
      ...attributes,
    },
  };
};

// Create a stripeAccount that conforms to the util/types stripeAccount schema
export const createStripeAccount = (id, attributes = {}) => ({
  id: new UUID(id),
  type: 'stripeAccount',
  attributes: {
    stripeAccountId: 'acc_testiaccountid',
    ...attributes,
  },
});

// Create a user that conforms to the util/types user schema
export const createUser = (id, attributes = {}) => ({
  id: new UUID(id),
  type: 'user',
  attributes: {
    banned: false,
    deleted: false,
    profile: {
      displayName: `${id} display name`,
      abbreviatedName: 'TT',
    },
    ...attributes,
  },
});

// Create a user that conforms to the util/types currentUser schema
export const createCurrentUser = (id, attributes = {}, includes = {}) => ({
  id: new UUID(id),
  type: 'currentUser',
  attributes: {
    banned: false,
    deleted: false,
    email: `${id}@example.com`,
    emailVerified: true,
    profile: {
      firstName: `${id} first name`,
      lastName: `${id} last name`,
      displayName: `${id} display name`,
      abbreviatedName: `${id} abbreviated name`,
      metadata: {
        isBrandAdmin: false,
      },
    },
    state: 'active',

    ...attributes,
  },
  effectivePermissionSet: {
    id: new UUID('test-effectivePermissionSet'),
    attributes: {
      initiateTransactions: 'permission/allow',
      postListings: 'permission/allow',
      read: 'permission/allow',
    },
  },
  ...includes,
});

// Create a user that conforms to the util/types user schema
export const createImage = id => ({
  id: new UUID(id),
  type: 'image',
  attributes: {
    variants: {
      // TODO: add all possible variants here
      square: {
        name: 'square',
        height: 408,
        width: 408,
        url: 'https://picsum.photos/408/408/',
      },
      square2x: {
        name: 'square2x',
        height: 816,
        width: 816,
        url: 'https://picsum.photos/816/816/',
      },
    },
  },
});

// Create a user that conforms to the util/types listing schema
export const createListing = (id, attributes = {}, includes = {}) => ({
  id: new UUID(id),
  type: 'listing',
  attributes: {
    title: `${id} title`,
    description: `${id} description`,
    geolocation: new LatLng(40, 60),
    deleted: false,
    state: LISTING_STATE_PUBLISHED,
    price: new Money(5500, 'USD'),
    availabilityPlan: {
      type: 'availability-plan/time',
      timezone: 'Etc/UTC',
    },
    publicData: {},
    ...attributes,
  },
  ...includes,
});

// Create a user that conforms to the util/types ownListing schema
export const createOwnListing = (id, attributes = {}, includes = {}) => ({
  id: new UUID(id),
  type: 'ownListing',
  attributes: {
    title: `${id} title`,
    description: `${id} description`,
    geolocation: new LatLng(40, 60),
    deleted: false,
    state: LISTING_STATE_PUBLISHED,
    price: new Money(5500, 'USD'),
    availabilityPlan: {
      type: 'availability-plan/time',
      timezone: 'Etc/UTC',
      entries: [
        { dayOfWeek: 'mon', startTime: '00:00', endTime: '00:00', seats: 1 },
        { dayOfWeek: 'tue', startTime: '00:00', endTime: '00:00', seats: 1 },
        { dayOfWeek: 'wed', startTime: '00:00', endTime: '00:00', seats: 1 },
        { dayOfWeek: 'thu', startTime: '00:00', endTime: '00:00', seats: 1 },
        { dayOfWeek: 'fri', startTime: '00:00', endTime: '00:00', seats: 1 },
        { dayOfWeek: 'sat', startTime: '00:00', endTime: '00:00', seats: 1 },
        { dayOfWeek: 'sun', startTime: '00:00', endTime: '00:00', seats: 1 },
      ],
    },
    publicData: {},
    ...attributes,
  },
  ...includes,
});

export const createStock = (id, attributes = {}) => {
  return {
    id: new UUID(id),
    type: 'stock',
    attributes: {
      quantity: 10,
      ...attributes,
    },
  };
};

export const createTxTransition = options => {
  return {
    createdAt: new Date(Date.UTC(2017, 4, 1)),
    by: TX_TRANSITION_ACTOR_CUSTOMER,
    transition: processTransitions.REQUEST_PAYMENT,
    ...options,
  };
};

export const createTransaction = options => {
  const {
    id,
    processName = 'default-purchase',
    processVersion = 1,
    lastTransition = processTransitions.CONFIRM_PAYMENT,
    total = new Money(1000, 'USD'),
    commission = new Money(100, 'USD'),
    booking = null,
    listing = null,
    customer = null,
    provider = null,
    reviews = [],
    lastTransitionedAt = new Date(Date.UTC(2017, 5, 1)),
    transitions = [
      createTxTransition({
        createdAt: new Date(Date.UTC(2017, 4, 1)),
        by: TX_TRANSITION_ACTOR_CUSTOMER,
        transition: processTransitions.REQUEST_PAYMENT,
      }),
      createTxTransition({
        createdAt: new Date(Date.UTC(2017, 4, 1, 0, 0, 1)),
        by: TX_TRANSITION_ACTOR_CUSTOMER,
        transition: processTransitions.CONFIRM_PAYMENT,
      }),
    ],
    lineItems: lineItemsParam,
  } = options;

  const dayCount = booking ? daysBetween(booking.attributes.start, booking.attributes.end) : 1;
  const lineItems = lineItemsParam
    ? lineItemsParam
    : [
        {
          code: 'line-item/item',
          includeFor: ['customer', 'provider'],
          quantity: new Decimal(dayCount),
          unitPrice: new Money(total.amount / dayCount, total.currency),
          lineTotal: total,
          reversal: false,
        },
        {
          code: 'line-item/provider-commission',
          includeFor: ['provider'],
          unitPrice: new Money(commission.amount * -1, commission.currency),
          lineTotal: new Money(commission.amount * -1, commission.currency),
          reversal: false,
        },
      ];

  return {
    id: new UUID(id),
    type: 'transaction',
    attributes: {
      createdAt: new Date(Date.UTC(2017, 4, 1)),
      processName,
      processVersion,
      lastTransitionedAt,
      lastTransition,
      payinTotal: total,
      payoutTotal: new Money(total.amount - commission.amount, total.currency),
      transitions,
      lineItems,
    },
    booking,
    listing,
    customer,
    provider,
    reviews,
  };
};

export const createMessage = (id, attributes = {}, includes = {}) => {
  return {
    id: new UUID(id),
    type: 'message',
    attributes: {
      createdAt: new Date(Date.UTC(2017, 10, 9, 8, 12)),
      content: `Message ${id}\n\nLorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.`,
      ...attributes,
    },
    ...includes,
  };
};

export const createReview = (id, attributes = {}, includes = {}) => {
  return {
    id: new UUID(id),
    attributes: {
      createdAt: new Date(),
      content:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.',
      rating: 3,
      state: 'public',
      type: 'ofProvider',
      ...attributes,
    },
    ...includes,
  };
};

/**
 * Creates an array of time slot objects.
 *
 * @param {Date} startDate date when the time slots start
 * @param {Number} numberOfDays number of time slots to create
 *
 * @return {Array} array of time slots
 */
export const createTimeSlots = (startDate, numberOfDays) => {
  return Array.from({ length: numberOfDays }, (v, i) => i).map(i => {
    return {
      id: new UUID(i),
      type: 'timeSlot',
      attributes: {
        start: getStartOf(startDate, 'day', 'Etc/UTC', i, 'days'),
        end: getStartOf(startDate, 'day', 'Etc/UTC', i + 1, 'days'),
        type: TIME_SLOT_TIME,
      },
    };
  });
};

// Default config for currency formatting in tests and examples.
export const currencyConfig = {
  style: 'currency',
  currency: 'USD',
  currencyDisplay: 'symbol',
  useGrouping: true,
  minimumFractionDigits: 2,
  maximumFractionDigits: 2,
};

const pad = num => {
  return num >= 0 && num < 10 ? `0${num}` : `${num}`;
};

// Create fake Internalization object to help with shallow rendering.
export const fakeIntl = {
  formatDate: d => `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())}`,
  formatDateTimeRange: (start, end) =>
    `${start.getUTCFullYear()}-${pad(start.getUTCMonth() + 1)}-${pad(
      start.getUTCDate()
    )} - ${end.getUTCFullYear()}-${pad(end.getUTCMonth() + 1)}-${pad(end.getUTCDate())}`,
  formatMessage: msg => msg.id,
  formatNumber: d => `${d}`,
  formatPlural: d => d,
  formatRelativeTime: d => d,
  formatTime: d => `${pad(d.getUTCHours())}:${pad(d.getUTCMinutes())}`,
  now: () => Date.UTC(2017, 10, 23, 12, 59),
  messages: {},
};

const noop = () => null;

export const fakeFormProps = {
  anyTouched: false,
  asyncValidating: false,
  dirty: false,
  form: 'fakeTestForm',
  invalid: false,
  pristine: true,
  clearSubmit: noop,
  touch: noop,
  untouch: noop,
  submit: noop,
  reset: noop,
  resetSection: noop,
  initialize: noop,
  handleSubmit: noop,
  destroy: noop,
  clearAsyncError: noop,
  clearFields: noop,
  clearSubmitErrors: noop,
  change: noop,
  blur: noop,
  autofill: noop,
  asyncValidate: noop,
  valid: true,
  submitSucceeded: false,
  submitFailed: false,
  submitting: false,
  pure: true,
  initialized: true,
};

// Create fake viewport to help with shallow rendering
export const fakeViewport = {
  width: 2100,
  height: 1339,
};
