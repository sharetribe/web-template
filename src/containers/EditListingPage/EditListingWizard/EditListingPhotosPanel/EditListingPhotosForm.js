import React, { useState } from 'react';
import { ARRAY_ERROR } from 'final-form';
import { Form as FinalForm, Field } from 'react-final-form';
import arrayMutators from 'final-form-arrays';
import { FieldArray } from 'react-final-form-arrays';
import isEqual from 'lodash/isEqual';
import classNames from 'classnames';

// Import configs and util modules
import { FormattedMessage, useIntl } from '../../../../util/reactIntl';
import { propTypes } from '../../../../util/types';
import { nonEmptyArray, composeValidators } from '../../../../util/validators';
import { isUploadImageOverLimitError } from '../../../../util/errors';

// Import shared components
import { Button, Form, AspectRatioWrapper } from '../../../../components';

// Import modules from this directory
import ListingImage from './ListingImage';
import css from './EditListingPhotosForm.module.css';

const ACCEPT_IMAGES = 'image/*';
const MAX_IMAGES = 10; // Maximum number of images allowed

const ImageUploadError = props => {
  return props.uploadOverLimit ? (
    <p className={css.error}>
      <FormattedMessage id="EditListingPhotosForm.imageUploadFailed.uploadOverLimit" />
    </p>
  ) : props.uploadImageError ? (
    <p className={css.error}>
      <FormattedMessage id="EditListingPhotosForm.imageUploadFailed.uploadFailed" />
    </p>
  ) : null;
};

// NOTE: PublishListingError and ShowListingsError are here since Photos panel is the last visible panel
// before creating a new listing. If that order is changed, these should be changed too.
// Create and show listing errors are shown above submit button
const PublishListingError = props => {
  return props.error ? (
    <p className={css.error}>
      <FormattedMessage id="EditListingPhotosForm.publishListingFailed" />
    </p>
  ) : null;
};

const ShowListingsError = props => {
  return props.error ? (
    <p className={css.error}>
      <FormattedMessage id="EditListingPhotosForm.showListingFailed" />
    </p>
  ) : null;
};

// Field component that uses file-input to allow user to select multiple images.
export const FieldAddImage = props => {
  const { formApi, onImageUploadHandler, aspectWidth = 1, aspectHeight = 1, currentImageCount = 0, ...rest } = props;
  return (
    <Field form={null} {...rest}>
      {fieldprops => {
        const { accept, input, label, disabled: fieldDisabled } = fieldprops;
        const { name, type } = input;
        const onChange = e => {
          const files = Array.from(e.target.files || []);
          if (files.length > 0) {
            // Check if adding these files would exceed the maximum limit
            const totalImages = currentImageCount + files.length;
            if (totalImages > MAX_IMAGES) {
              // Show error or truncate to max allowed
              const allowedFiles = files.slice(0, MAX_IMAGES - currentImageCount);
              onImageUploadHandler(allowedFiles);
            } else {
              onImageUploadHandler(files);
            }
          }
          // Clear the input value to allow selecting the same files again
          e.target.value = '';
        };
        const inputProps = { accept, id: name, name, onChange, type: 'file', multiple: true };
        const isDisabled = fieldDisabled || currentImageCount >= MAX_IMAGES;
        
        return (
          <div className={css.addImageWrapper}>
            <AspectRatioWrapper width={aspectWidth} height={aspectHeight}>
              {isDisabled ? null : <input {...inputProps} className={css.addImageInput} />}
              <label htmlFor={name} className={classNames(css.addImage, { [css.disabled]: isDisabled })}>
                {label}
              </label>
            </AspectRatioWrapper>
          </div>
        );
      }}
    </Field>
  );
};

// Component that shows listing images from "images" field array
const FieldListingImage = props => {
  const { name, intl, onRemoveImage, aspectWidth, aspectHeight, variantPrefix } = props;
  return (
    <Field name={name}>
      {fieldProps => {
        const { input } = fieldProps;
        const image = input.value;
        return image ? (
          <ListingImage
            image={image}
            key={image?.id?.uuid || image?.id}
            className={css.thumbnail}
            savedImageAltText={intl.formatMessage({
              id: 'EditListingPhotosForm.savedImageAltText',
            })}
            onRemoveImage={() => onRemoveImage(image?.id)}
            aspectWidth={aspectWidth}
            aspectHeight={aspectHeight}
            variantPrefix={variantPrefix}
          />
        ) : null;
      }}
    </Field>
  );
};

/**
 * The EditListingPhotosForm component.
 *
 * @component
 * @param {Object} props
 * @param {string} [props.className] - Custom class that extends the default class for the root element
 * @param {string} [props.rootClassName] - Custom class that overrides the default class for the root element
 * @param {boolean} props.disabled - Whether the form is disabled
 * @param {boolean} props.ready - Whether the form is ready
 * @param {boolean} props.updated - Whether the form is updated
 * @param {boolean} props.updateInProgress - Whether the update is in progress
 * @param {Object} props.fetchErrors - The fetch errors object
 * @param {propTypes.error} props.fetchErrors.publishListingError - The publish listing error
 * @param {propTypes.error} props.fetchErrors.showListingsError - The show listings error
 * @param {propTypes.error} props.fetchErrors.uploadImageError - The upload image error
 * @param {propTypes.error} props.fetchErrors.updateListingError - The update listing error
 * @param {string} props.saveActionMsg - The save action message
 * @param {Function} props.onSubmit - The submit function
 * @param {Function} props.onImageUpload - The image upload function
 * @param {Function} props.onRemoveImage - The remove image function
 * @param {Object} props.listingImageConfig - The listing image config
 * @param {number} props.listingImageConfig.aspectWidth - The aspect width
 * @param {number} props.listingImageConfig.aspectHeight - The aspect height
 * @param {string} props.listingImageConfig.variantPrefix - The variant prefix
 * @returns {JSX.Element}
 */
export const EditListingPhotosForm = props => {
  const [state, setState] = useState({ imageUploadRequested: false, uploadingCount: 0 });
  const [submittedImages, setSubmittedImages] = useState([]);

  const onImageUploadHandler = files => {
    const { listingImageConfig, onImageUpload } = props;
    if (files && files.length > 0) {
      setState(prevState => ({ 
        imageUploadRequested: true, 
        uploadingCount: prevState.uploadingCount + files.length 
      }));

      // Upload files sequentially to avoid overwhelming the server
      const uploadPromises = files.map(file => {
        const imagePayload = { id: `${file.name}_${Date.now()}_${Math.random()}`, file };
        return onImageUpload(imagePayload, listingImageConfig)
          .then(() => {
            setState(prevState => ({ 
              ...prevState, 
              uploadingCount: prevState.uploadingCount - 1 
            }));
          })
          .catch(() => {
            setState(prevState => ({ 
              ...prevState, 
              uploadingCount: prevState.uploadingCount - 1 
            }));
          });
      });

      Promise.all(uploadPromises).finally(() => {
        setState(prevState => ({ 
          ...prevState, 
          imageUploadRequested: prevState.uploadingCount > 1 
        }));
      });
    }
  };

  const intl = useIntl();

  return (
    <FinalForm
      {...props}
      mutators={{ ...arrayMutators }}
      render={formRenderProps => {
        const {
          form,
          className,
          fetchErrors,
          handleSubmit,
          invalid,
          onRemoveImage,
          disabled,
          ready,
          saveActionMsg,
          updated,
          updateInProgress,
          touched,
          errors,
          values,
          listingImageConfig,
        } = formRenderProps;

        const images = values.images || [];
        const { aspectWidth = 1, aspectHeight = 1, variantPrefix } = listingImageConfig;

        const { publishListingError, showListingsError, updateListingError, uploadImageError } =
          fetchErrors || {};
        const uploadOverLimit = isUploadImageOverLimitError(uploadImageError);

        // imgs can contain added images (with temp ids) and submitted images with uniq ids.
        const arrayOfImgIds = imgs => imgs?.map(i => (typeof i.id === 'string' ? i.imageId : i.id));
        const imageIdsFromProps = arrayOfImgIds(images);
        const imageIdsFromPreviousSubmit = arrayOfImgIds(submittedImages);
        const imageArrayHasSameImages = isEqual(imageIdsFromProps, imageIdsFromPreviousSubmit);
        const submittedOnce = submittedImages.length > 0;
        const pristineSinceLastSubmit = submittedOnce && imageArrayHasSameImages;

        const submitReady = (updated && pristineSinceLastSubmit) || ready;
        const submitInProgress = updateInProgress;
        const submitDisabled =
          invalid || disabled || submitInProgress || state.imageUploadRequested || ready;
        const imagesError = touched.images && errors?.images && errors.images[ARRAY_ERROR];

        const classes = classNames(css.root, className);

        return (
          <Form
            className={classes}
            onSubmit={e => {
              setSubmittedImages(images);
              handleSubmit(e);
            }}
          >
            {updateListingError ? (
              <p className={css.error}>
                <FormattedMessage id="EditListingPhotosForm.updateFailed" />
              </p>
            ) : null}

            <div className={css.imagesFieldArray}>
              <FieldArray
                name="images"
                validate={composeValidators(
                  nonEmptyArray(
                    intl.formatMessage({
                      id: 'EditListingPhotosForm.imageRequired',
                    })
                  )
                )}
              >
                {({ fields }) =>
                  fields.map((name, index) => (
                    <FieldListingImage
                      key={name}
                      name={name}
                      onRemoveImage={imageId => {
                        fields.remove(index);
                        onRemoveImage(imageId);
                      }}
                      intl={intl}
                      aspectWidth={aspectWidth}
                      aspectHeight={aspectHeight}
                      variantPrefix={variantPrefix}
                    />
                  ))
                }
              </FieldArray>

              <FieldAddImage
                id="addImage"
                name="addImage"
                accept={ACCEPT_IMAGES}
                label={
                  <span className={css.chooseImageText}>
                    <span className={css.chooseImage}>
                      <FormattedMessage id="EditListingPhotosForm.chooseImages" />
                    </span>
                    <span className={css.imageTypes}>
                      <FormattedMessage id="EditListingPhotosForm.imageTypes" />
                    </span>
                    {images.length > 0 && (
                      <span className={css.imageCount}>
                        <FormattedMessage 
                          id="EditListingPhotosForm.imageCount" 
                          values={{ count: images.length, max: MAX_IMAGES }}
                        />
                      </span>
                    )}
                  </span>
                }
                type="file"
                disabled={state.imageUploadRequested}
                formApi={form}
                onImageUploadHandler={onImageUploadHandler}
                aspectWidth={aspectWidth}
                aspectHeight={aspectHeight}
                currentImageCount={images.length}
              />
            </div>

            {imagesError ? <div className={css.arrayError}>{imagesError}</div> : null}

            <ImageUploadError
              uploadOverLimit={uploadOverLimit}
              uploadImageError={uploadImageError}
            />

            {state.uploadingCount > 0 && (
              <p className={css.uploadingMessage}>
                <FormattedMessage 
                  id="EditListingPhotosForm.uploadingImages" 
                  values={{ count: state.uploadingCount }}
                />
              </p>
            )}

            <p className={css.tip}>
              <FormattedMessage id="EditListingPhotosForm.addImagesTip" />
            </p>

            <PublishListingError error={publishListingError} />
            <ShowListingsError error={showListingsError} />

            <Button
              className={css.submitButton}
              type="submit"
              inProgress={submitInProgress}
              disabled={submitDisabled}
              ready={submitReady}
            >
              {saveActionMsg}
            </Button>
          </Form>
        );
      }}
    />
  );
};

export default EditListingPhotosForm;
