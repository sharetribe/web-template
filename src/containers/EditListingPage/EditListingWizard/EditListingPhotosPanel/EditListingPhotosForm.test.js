import React from 'react';
import '@testing-library/jest-dom';
import { Form as FinalForm, Field } from 'react-final-form';
import arrayMutators from 'final-form-arrays';

import { fakeIntl } from '../../../../util/testData';
import { renderWithProviders as render, testingLibrary } from '../../../../util/testHelpers';

import EditListingPhotosForm, { FieldAddImage } from './EditListingPhotosForm';

const { screen, userEvent, waitFor, act } = testingLibrary;

const noop = () => null;

describe('EditListingPhotosForm', () => {
  it('matches snapshot', () => {
    const saveActionMsg = 'Save photos';
    const tree = render(
      <EditListingPhotosForm
        initialValues={{ country: 'US', images: [] }}
        intl={fakeIntl}
        dispatch={noop}
        onImageUpload={v => Promise.reject(v)}
        onSubmit={v => v}
        saveActionMsg={saveActionMsg}
        stripeConnected={false}
        updated={false}
        ready={false}
        updateInProgress={false}
        disabled={false}
        onRemoveImage={noop}
        listingImageConfig={{ aspectWidth: 1, aspectHeight: 1, variantPrefix: 'listing-card' }}
      />
    );
    expect(tree.asFragment()).toMatchSnapshot();
  });

  // TODO to test this fully, we would need to check that store's state changes correctly.

  it('Check that FieldAddImage works with single file', async () => {
    const ACCEPT_IMAGES = 'image/*';
    const tree = render(
      <FinalForm
        onSubmit={noop}
        mutators={{ ...arrayMutators }}
        render={formRenderProps => {
          return (
            <form onSubmit={noop}>
              <FieldAddImage
                id="addImage"
                name="addImage"
                accept={ACCEPT_IMAGES}
                label={<div>label</div>}
                type="file"
                disabled={false}
                formApi={{
                  change: noop,
                  blur: noop,
                }}
                onImageUploadHandler={noop}
                aspectWidth={1}
                aspectHeight={1}
                currentImageCount={0}
              />
            </form>
          );
        }}
      />
    );

    // Fill mandatory attributes
    const file = new File(['hello'], './public/static/icons/favicon-16x16.png', {
      type: 'image/png',
    });
    const input = screen.getByLabelText(/label/i);

    userEvent.upload(input, file);
    expect(input.files[0]).toBe(file);
    expect(input.files.item(0)).toBe(file);
    expect(input.files).toHaveLength(1);
  });

  it('Check that FieldAddImage works with multiple files', async () => {
    const ACCEPT_IMAGES = 'image/*';
    const onImageUploadHandler = jest.fn();
    
    const tree = render(
      <FinalForm
        onSubmit={noop}
        mutators={{ ...arrayMutators }}
        render={formRenderProps => {
          return (
            <form onSubmit={noop}>
              <FieldAddImage
                id="addImage"
                name="addImage"
                accept={ACCEPT_IMAGES}
                label={<div>label</div>}
                type="file"
                disabled={false}
                formApi={{
                  change: noop,
                  blur: noop,
                }}
                onImageUploadHandler={onImageUploadHandler}
                aspectWidth={1}
                aspectHeight={1}
                currentImageCount={0}
              />
            </form>
          );
        }}
      />
    );

    // Create multiple files
    const file1 = new File(['hello1'], 'image1.png', { type: 'image/png' });
    const file2 = new File(['hello2'], 'image2.png', { type: 'image/png' });
    const file3 = new File(['hello3'], 'image3.png', { type: 'image/png' });
    
    const input = screen.getByLabelText(/label/i);

    userEvent.upload(input, [file1, file2, file3]);
    
    expect(input.files).toHaveLength(3);
    expect(input.files[0]).toBe(file1);
    expect(input.files[1]).toBe(file2);
    expect(input.files[2]).toBe(file3);
    
    // Check that onImageUploadHandler was called with the array of files
    expect(onImageUploadHandler).toHaveBeenCalledWith([file1, file2, file3]);
  });

  it('Check that FieldAddImage respects maximum image limit', async () => {
    const ACCEPT_IMAGES = 'image/*';
    const onImageUploadHandler = jest.fn();
    
    const tree = render(
      <FinalForm
        onSubmit={noop}
        mutators={{ ...arrayMutators }}
        render={formRenderProps => {
          return (
            <form onSubmit={noop}>
              <FieldAddImage
                id="addImage"
                name="addImage"
                accept={ACCEPT_IMAGES}
                label={<div>label</div>}
                type="file"
                disabled={false}
                formApi={{
                  change: noop,
                  blur: noop,
                }}
                onImageUploadHandler={onImageUploadHandler}
                aspectWidth={1}
                aspectHeight={1}
                currentImageCount={8} // Already have 8 images
              />
            </form>
          );
        }}
      />
    );

    // Create multiple files that would exceed the limit
    const files = Array.from({ length: 5 }, (_, i) => 
      new File([`hello${i}`], `image${i}.png`, { type: 'image/png' })
    );
    
    const input = screen.getByLabelText(/label/i);

    userEvent.upload(input, files);
    
    // Should only allow 2 more files (10 - 8 = 2)
    expect(onImageUploadHandler).toHaveBeenCalledWith(files.slice(0, 2));
  });

  it('Check that FieldAddImage is disabled when max images reached', async () => {
    const ACCEPT_IMAGES = 'image/*';
    const onImageUploadHandler = jest.fn();
    
    const tree = render(
      <FinalForm
        onSubmit={noop}
        mutators={{ ...arrayMutators }}
        render={formRenderProps => {
          return (
            <form onSubmit={noop}>
              <FieldAddImage
                id="addImage"
                name="addImage"
                accept={ACCEPT_IMAGES}
                label={<div>label</div>}
                type="file"
                disabled={false}
                formApi={{
                  change: noop,
                  blur: noop,
                }}
                onImageUploadHandler={onImageUploadHandler}
                aspectWidth={1}
                aspectHeight={1}
                currentImageCount={10} // Max images reached
              />
            </form>
          );
        }}
      />
    );

    // When max images are reached, the input is not rendered and the label has disabled class
    const label = screen.getByLabelText(/label/i);
    expect(label).toHaveClass('disabled');
  });
});
