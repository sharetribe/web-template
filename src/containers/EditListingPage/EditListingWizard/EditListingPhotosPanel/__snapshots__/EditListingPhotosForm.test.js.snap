// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditListingDeliveryForm matches snapshot 1`] = `
<DocumentFragment>
  <form
    action="/"
    class="root"
    method="post"
  >
    <div
      class="imagesFieldArray"
    >
      <div
        class="addImageWrapper"
      >
        <div
          class="root"
        >
          <div
            class="aspectPadding"
            style="padding-bottom: 100%;"
          >
            <div
              class="aspectBox"
            >
              <input
                accept="image/*"
                class="addImageInput"
                id="addImage"
                name="addImage"
                type="file"
              />
              <label
                class="addImage"
                for="addImage"
              >
                <span
                  class="chooseImageText"
                >
                  <span
                    class="chooseImage"
                  >
                    <span>
                      EditListingPhotosForm.chooseImage
                    </span>
                  </span>
                  <span
                    class="imageTypes"
                  >
                    <span>
                      EditListingPhotosForm.imageTypes
                    </span>
                  </span>
                </span>
              </label>
            </div>
          </div>
        </div>
      </div>
    </div>
    <p
      class="tip"
    >
      <span>
        EditListingPhotosForm.addImagesTip
      </span>
    </p>
    <button
      class="root submitButton"
      disabled=""
      type="submit"
    >
      Save photos
    </button>
  </form>
</DocumentFragment>
`;

exports[`EditListingPhotosForm matches snapshot 1`] = `
<DocumentFragment>
  <form
    action="/"
    class="root"
    method="post"
  >
    <div
      class="imagesFieldArray"
    >
      <div
        class="addImageWrapper"
      >
        <div
          class="root"
        >
          <div
            class="aspectPadding"
            style="padding-bottom: 100%;"
          >
            <div
              class="aspectBox"
            >
              <input
                accept="image/*"
                class="addImageInput"
                id="addImage"
                multiple=""
                name="addImage"
                type="file"
              />
              <label
                class="addImage"
                for="addImage"
              >
                <span
                  class="chooseImageText"
                >
                  <span
                    class="chooseImage"
                  >
                    <span>
                      EditListingPhotosForm.chooseImages
                    </span>
                  </span>
                  <span
                    class="imageTypes"
                  >
                    <span>
                      EditListingPhotosForm.imageTypes
                    </span>
                  </span>
                </span>
              </label>
            </div>
          </div>
        </div>
      </div>
    </div>
    <p
      class="tip"
    >
      <span>
        EditListingPhotosForm.addImagesTip
      </span>
    </p>
    <button
      class="root submitButton"
      disabled=""
      type="submit"
    >
      Save photos
    </button>
  </form>
</DocumentFragment>
`;
